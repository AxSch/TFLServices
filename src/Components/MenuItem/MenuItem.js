import React from 'react'
import { useDispatch } from 'react-redux'
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome'
import { setCurrentTFLService, setIsSearch } from '../../reducers/transportServices/transportServicesSlice'
import { Item, DisruptionIcons } from './MenuItem.styled'

const MenuItem = ({ service }) => {
    const dispatch = useDispatch()

    const renderNightIcon = (serviceTypes) => {
        const nightService = serviceTypes.filter(serviceType => serviceType.name === 'Night')
        if (nightService.length > 0) {
            return <FontAwesomeIcon data-testid="nightServiceIcon" icon="cloud-moon" color={'#141852'} />
        }
    }

    const renderDisruptionIcon = (lineStatus) => {
        const lineStatuses = lineStatus.filter(line => line.statusSeverity !== 10)
        let statusSeverity
        if (lineStatuses.length > 0) {
            statusSeverity = lineStatuses.map(line => {
                return line.statusSeverityDescription
            })
        } else {
            statusSeverity = lineStatus.map(line => {
                return line.statusSeverityDescription
            })
        }
        if (statusSeverity.length === 1 && statusSeverity[0] === 'Good Service') {
            return <FontAwesomeIcon data-testid="goodSeviceIcon" icon="check-circle" color={'#4BB543'}/>
        } else {
            return <FontAwesomeIcon data-testid="disruptionsIcon" icon="exclamation-circle" color={'#F32013'} />
        }
    }

    const handleOnClick = () => {
        dispatch(setCurrentTFLService(service))
        dispatch(setIsSearch(false))
    }

    return (
        <Item onClick={() => handleOnClick()} data-testid="menuItem">
            <div>
                {service.name}
            </div>
            <DisruptionIcons>
                {renderDisruptionIcon(service.lineStatuses)}
                {renderNightIcon(service.serviceTypes)}
            </DisruptionIcons>
        </Item>
    )
}


export default MenuItem
