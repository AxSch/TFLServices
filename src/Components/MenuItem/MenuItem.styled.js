import styled from 'styled-components'

const Item = styled.li`
    padding: 0.7rem 0.5rem;
    cursor: pointer;
    border-bottom: 1px solid #d8d8d8;
    
    &:hover {
        background-color: #ececec;
        border-radius: 4px;
    }

    &:last-child {
        border-bottom: none;
    }
`

const DisruptionIcons = styled.div`
    display: flex;
    align-items: center;
    padding: 0.4rem 0;

    svg {
        margin-right: 0.4rem;
    }
`

export {
    Item,
    DisruptionIcons,
}
