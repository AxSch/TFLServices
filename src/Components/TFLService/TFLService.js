import React, { useState, useEffect } from 'react'

const TFLService = ({ service }) => {
    const [isDisrupted, setIsDisrupted] = useState(false)
    const severeLineStatuses = service.lineStatuses.filter(line => line.statusSeverity !== 10)

    useEffect(() => {
        if (severeLineStatuses.length > 0) {
            if (severeLineStatuses[0].statusSeverity !== 10 ) {
                setIsDisrupted(true)
            }
        } else {
            setIsDisrupted(false)
        }
    }, [isDisrupted, severeLineStatuses])

    const renderDisruptionHeader = () => {
        if (severeLineStatuses.length > 0) {
            return (
                <span>
                    Service is currently suffering disruptions
                </span>
            )
        } else {
            return (
                <span>
                    No service disruptions
                </span>
            )
        }
    }

    const renderDisruptionsList = () => {
        if (severeLineStatuses.length > 0) {
            return severeLineStatuses.map((line, index) => {
                return (
                    <li key={index}>
                        {line.reason}
                    </li>
                )
            })
        }
    }
    const renderDisruptions = () => {
        if (isDisrupted) {
            return (
                <div>
                    <h3>Disruptions:</h3>
                    <ul>
                        {renderDisruptionsList()}
                    </ul>
                </div>
            )
        }
    }
    // console.log(isDisrupted)
    return (
        <div>
            <div>
                <h2>{service.name}</h2>
                <span>{renderDisruptionHeader()}</span>
            </div>
            {renderDisruptions()}
        </div>
    )
}

export default TFLService
