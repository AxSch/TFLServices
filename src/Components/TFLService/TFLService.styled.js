import styled from 'styled-components'

const TFLServiceContainer = styled.div`
    margin-top: 1.2rem;
    display: flex;
    justify-content: center;
    width: 100%;
    flex-wrap: wrap;
    padding: 1rem;

    @media(min-width: 768px) {
        margin-top: 0;
        align-content: start;
    }
    
    @media(min-width: 1024px) {
        justify-content: start;
        width: 50%;
    }
`
const ServiceHeader = styled.div`
    display: flex;
    flex-direction: column;

    @media(min-width: 1024px) {
        align-items: start;
    }
`

const ServiceDisruptions = styled.div`
    display: flex;
    flex-wrap: wrap;
    justify-content: center;

    @media(min-width: 1024px) {
        align-content: start;
        justify-content: start;
    }
`

const ServiceDisruptionsList = styled.ul`
    list-style-type: none;
    text-align: left;
    padding-left: 1.4rem;

    @media(min-width: 1024px) {
        padding-left: 0; 
    }
`

const DisruptionsListItem = styled.li`
    margin: 1.4rem 0;
`


export {
    ServiceHeader,
    TFLServiceContainer,
    ServiceDisruptions,
    ServiceDisruptionsList,
    DisruptionsListItem,
}
