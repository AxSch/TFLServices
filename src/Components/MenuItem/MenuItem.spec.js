import React from 'react'
import renderConnected from '../../utils/testHelpers';
import MenuItem from './MenuItem'
import { mockService } from '../TFLService/TFLService.spec'
import registerIcons from '../../App'

const initialState = {
    isSearch: false,
    previousBikeResults: {},
    currentTFLService: {}
}

describe('MenuItem', () => {
    let container, getByTestId, queryByTestId
    registerIcons()
    
    it('should render successfully',  () => {
        const utils = renderConnected(<MenuItem service={mockService} />, { initialState });
        container = utils.container;
        getByTestId = utils.getByTestId;
        expect(getByTestId("menuItem")).toBeInTheDocument();
        expect(container).toMatchSnapshot();
    })

    it('should render the night icon - Night travel provided',  () => {
        const utils = renderConnected(<MenuItem service={mockService} />, { initialState });
        container = utils.container;
        getByTestId = utils.getByTestId;
        expect(getByTestId("nightServiceIcon")).toBeInTheDocument();
    })

    it('should render the disruption icon - disruptions on the service',  () => {
        const utils = renderConnected(<MenuItem service={mockService} />, { initialState });
        container = utils.container;
        getByTestId = utils.getByTestId;
        expect(getByTestId("disruptionsIcon")).toBeInTheDocument();
    })
    
    it('should not render the disruption icon - service is good',  () => {
        mockService.lineStatuses = [
            {
                $type: "Tfl.Api.Presentation.Entities.LineStatus, Tfl.Api.Presentation.Entities",
                created: "0001-01-01T00:00:00",
                id: 0,
                statusSeverity: 10,
                statusSeverityDescription: "Good Service"
            }
        ]
        const utils = renderConnected(<MenuItem service={mockService} />, { initialState });
        container = utils.container;
        queryByTestId = utils.queryByTestId;
        expect(queryByTestId("disruptionsIcon")).not.toBeInTheDocument();
        expect(queryByTestId("goodSeviceIcon")).toBeInTheDocument();
    })

    it('should not render the night icon - night service is not enabled',  () => {
        mockService.serviceTypes = [
            {
                $type: "Tfl.Api.Presentation.Entities.LineServiceTypeInfo, Tfl.Api.Presentation.Entities",
                name: "Regular",
                uri: "/Line/Route?ids=DLR&serviceTypes=Regular"
            }
        ]
        const utils = renderConnected(<MenuItem service={mockService} />, { initialState });
        container = utils.container;
        queryByTestId = utils.queryByTestId;
        expect(queryByTestId("nightServiceIcon")).not.toBeInTheDocument();
    })
})
