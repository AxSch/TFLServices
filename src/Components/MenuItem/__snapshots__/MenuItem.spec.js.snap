// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MenuItem should render successfully 1`] = `
<div>
  <li
    class="sc-bdnxRM kQNWVX"
    data-testid="menuItem"
  >
    <div>
      London Overground
    </div>
    <div
      class="sc-gtsrHT dlgqIE"
    >
      <svg
        aria-hidden="true"
        class="svg-inline--fa fa-exclamation-circle fa-w-16 "
        color="#F32013"
        data-icon="exclamation-circle"
        data-prefix="fas"
        data-testid="disruptionsIcon"
        focusable="false"
        role="img"
        viewBox="0 0 512 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M504 256c0 136.997-111.043 248-248 248S8 392.997 8 256C8 119.083 119.043 8 256 8s248 111.083 248 248zm-248 50c-25.405 0-46 20.595-46 46s20.595 46 46 46 46-20.595 46-46-20.595-46-46-46zm-43.673-165.346l7.418 136c.347 6.364 5.609 11.346 11.982 11.346h48.546c6.373 0 11.635-4.982 11.982-11.346l7.418-136c.375-6.874-5.098-12.654-11.982-12.654h-63.383c-6.884 0-12.356 5.78-11.981 12.654z"
          fill="currentColor"
        />
      </svg>
      <svg
        aria-hidden="true"
        class="svg-inline--fa fa-cloud-moon fa-w-18 "
        color="#141852"
        data-icon="cloud-moon"
        data-prefix="fas"
        data-testid="nightServiceIcon"
        focusable="false"
        role="img"
        viewBox="0 0 576 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M342.8 352.7c5.7-9.6 9.2-20.7 9.2-32.7 0-35.3-28.7-64-64-64-17.2 0-32.8 6.9-44.3 17.9-16.3-29.6-47.5-49.9-83.7-49.9-53 0-96 43-96 96 0 2 .5 3.8.6 5.7C27.1 338.8 0 374.1 0 416c0 53 43 96 96 96h240c44.2 0 80-35.8 80-80 0-41.9-32.3-75.8-73.2-79.3zm222.5-54.3c-93.1 17.7-178.5-53.7-178.5-147.7 0-54.2 29-104 76.1-130.8 7.3-4.1 5.4-15.1-2.8-16.7C448.4 1.1 436.7 0 425 0 319.1 0 233.1 85.9 233.1 192c0 8.5.7 16.8 1.8 25 5.9 4.3 11.6 8.9 16.7 14.2 11.4-4.7 23.7-7.2 36.4-7.2 52.9 0 96 43.1 96 96 0 3.6-.2 7.2-.6 10.7 23.6 10.8 42.4 29.5 53.5 52.6 54.4-3.4 103.7-29.3 137.1-70.4 5.3-6.5-.5-16.1-8.7-14.5z"
          fill="currentColor"
        />
      </svg>
    </div>
  </li>
</div>
`;

exports[`TFLService should render successfully 1`] = `
<div>
  <div
    class="sc-dlnjwi eDwYie"
    data-testid="tflService"
  >
    <div
      class="sc-hKFxyN eAVAki"
    >
      <h2>
        London Overground
      </h2>
      <span
        data-testid="hasDisruptions"
      >
        Service is currently suffering disruptions
      </span>
    </div>
    <div
      class="sc-eCApnc bJhYwM"
    >
      <h3>
        Disruptions:
      </h3>
      <ul
        class="sc-jSFjdj erazVT"
        data-testid="disruptionsList"
      >
        <li
          class="sc-gKAaRy jjKBix"
        >
          Changes to Stratford to Richmond services on Mondays to Fridays from Monday 17 May until Friday 18 June: The following trains from Stratford to Richmond will not run: 06:05, 07:18, 08:30, 09:40, 10:59, 12:14, 13:29, 14:43, 15:52, 17:05, 18:16, 19:29, 20:44, 21:59 and 23:14. The 22:59 is now the last train from Stratford for stations from Willesden Junction to Richmond and Clapham Junction - later trains will run from Stratford as far as Willesden Junction only. The following trains from Richmond to Stratford will not run: 06:08. 07:21, 08:34, 09:47, 11:01, 12:17, 13:31, 14:45, 15:56, 17:08, 18:20, 19:32, 20:47, 22:01 and 23:17
        </li>
      </ul>
    </div>
  </div>
</div>
`;
